wrangler-service/src/main/java/io/cdap/wrangler/service/directive/RemoteExecutionTask.java

```diff
@@ -19,43 +19,109 @@
   public void run(RunnableTaskContext runnableTaskContext) throws Exception {
-    RemoteDirectiveRequest directiveRequest = GSON
-      .fromJson(runnableTaskContext.getParam(), RemoteDirectiveRequest.class);
+    RemoteDirectiveRequest directiveRequest = GSON.fromJson(runnableTaskContext.getParam(),
+                                                            RemoteDirectiveRequest.class);
+
     SystemAppTaskContext systemAppContext = runnableTaskContext.getRunnableTaskSystemAppContext();
     String namespace = directiveRequest.getPluginNameSpace();
-    try (DirectiveRegistry composite = new CompositeDirectiveRegistry(
-      new SystemDirectiveRegistry(),
-      new UserDirectiveRegistry(systemAppContext))) {
-      composite.reload(namespace);
-      Function<TransientStore, ExecutorContext> contextProvider =
-        store -> new ServicePipelineContext(namespace, ExecutorContext.Environment.SERVICE, systemAppContext, store);
-      CommonDirectiveExecutor commonDirectiveExecutor = new CommonDirectiveExecutor(contextProvider, composite);
+    Map<String, DirectiveClass> systemDirectives = directiveRequest.getSystemDirectives();
+
+    // Collect directives.
+    try (UserDirectiveRegistry userDirectiveRegistry = new UserDirectiveRegistry(systemAppContext)) {
+      List<Directive> directives = new ArrayList<>();
+      GrammarWalker walker = new GrammarWalker(new RecipeCompiler(), new ConfigDirectiveContext(DirectiveConfig.EMPTY));
+      walker.walk(directiveRequest.getRecipe(), (command, tokenGroup) -> {
+        DirectiveInfo info;
+        DirectiveClass directiveClass = systemDirectives.get(command);
+        if (directiveClass == null) {
+          info = userDirectiveRegistry.get(namespace, command);
+        } else {
+          // For system directives, we can load it directly from the classloader.
+          try {
+            info = DirectiveInfo.fromSystem((Class<? extends Directive>) Class.forName(directiveClass.getClassName()));
+          } catch (ClassNotFoundException e) {
+            throw new DirectiveLoadException("Failed to load system directive " + directiveClass.getName(), e);
+          }
+        }
+
+        Directive directive = info.instance();
+        UsageDefinition definition = directive.define();
+        Arguments arguments = new MapArguments(definition, tokenGroup);
+        directive.initialize(arguments);
+        directives.add(directive);
+      });
+
+      // If there is no directives, there is nothing to execute
+      if (directives.isEmpty()) {
+        runnableTaskContext.writeResult(directiveRequest.getData());
+        return;
+      }
+
       ObjectSerDe<List<Row>> objectSerDe = new ObjectSerDe<>();
-      List<Row> inputRows = objectSerDe.toObject(directiveRequest.getData());
-      List<Row> rows = commonDirectiveExecutor
-        .executeDirectives(namespace, directiveRequest.getDirectives(), inputRows);
+      List<Row> rows = objectSerDe.toObject(directiveRequest.getData());
```

Comment:
I am wondering if we should just use https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/SerializationUtils.html

Reply:
Firstly, I was not intended to change existing logic. Secondly, based on actual result, it seems not too slow. Lastly, I am a bit reluctant to use apache common library.